pyrevolve/pyrevolve.py

```diff
@@ -115,107 +212,354 @@ def apply_reverse(self):
         recompute sections of the trajectory that have not been stored in the
         forward run."""
 
-        with self.profiler.get_timer('reverse', 'reverse'):
-            self.rev_operator.apply(t_start=self.scheduler.capo,
-                                    t_end=self.scheduler.capo+1)
-        while(True):
+        action = None
+        remove_ckp_flag = False
+        while True:
             # ask Revolve what to do next.
-            action = self.scheduler.next()
-            if(action.type == Action.ADVANCE):
-                # advance forward computation
-                with self.profiler.get_timer('reverse', 'advance'):
-                    self.fwd_operator.apply(t_start=self.scheduler.old_capo,
-                                            t_end=self.scheduler.capo)
-            elif(action.type == Action.TAKESHOT):
+            if remove_ckp_flag is True:
+                action = HAction(
+                    action_type=Action.CPDEL,
+                    capo=self.__last_capo_read,
+                    index=[self.__last_stidx_read],
+                    old_capo=self.scheduler.old_capo,
+                    ckp=self.scheduler.cp_pointer,
+                )
+                remove_ckp_flag = False
+            else:
+                action = self.scheduler.next()
+
+            if action.type == Action.REVERSE:
+                # advance adjoint computation by a single step
+                with self.profiler.get_timer("reverse", "reverse"):
+                    self.fwd_operator.apply(
+                        t_start=self.scheduler.capo, t_end=self.scheduler.capo + 1
+                    )
+                    self.rev_operator.apply(
+                        t_start=self.scheduler.capo, t_end=self.scheduler.capo + 1
+                    )
+                if self.scheduler.capo == self.__last_capo_read:
+                    # h-revolve implies that B^i must also remove
+                    # the checkpoint if it was previously restored
+                    remove_ckp_flag = True
```

Comment:
Can this logic not move to schedulers/hrevolve.py so that by this point there is a CPDEL Action in the provided schedule?

Reply:
I think it would be hard to move it for hrevolve.py because I would have to mess with the original h-revolve codes. I can easily move it to schedulers/__init__.py, where the HRevolve class is defined. What do you think?

Reply:
Yes please. This belongs in the `HRevolve` class.

Reply:
Consider it done! :-)